### PR TITLE
Add Qt Tree Model and View

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR builds on (and includes) #2217, adding a complementary `QtAbstractItemModel` and a `QTreeView` that can follow and/or control a `NestedEventedList[Node]`.  Ultimately, for LayerGroups (#970), `QtLayerList` would become a `QTreeView`, with a special [`QItemDelegate`](https://doc.qt.io/qt-5/qitemdelegate.html) to display the layers.

#2217 should be reviewed and merged first.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by goanpeca** [comment]
Any image/screenshot associated with this PR 🙃 ?

**Comment by tlambert03** [comment]
> Any image/screenshot associated with this PR 🙃 ?

it's not too much to look at at this point.  it's just an unstyled model.  You can play with it at `python examples/node_tree.py`

Untitled

**Comment by tlambert03** [comment]
@sofroniewn, @jni, this is also pretty well tested and ready for review.  Do #2217 first, as this contains that and will be easier to review once those changes are not visible here.  Again, nothing is actually hooked up here, but there is an example that you can run if you'd like to see the general treeview/model interaction

File: examples/node_tree.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+end = "\033[0m"
+Bold = "\033[1m"
+Dim = "\033[2m"
+ResetDim = "\033[22m"
+red = "\033[0;31m"
+green = "\033[0;32m"
+colorlog_format = f'{green}%(levelname)9s:{end} {Dim}%(name)36s.{ResetDim}{red}%(funcName)-18s{end}{"%(message)s"}'
+logging.basicConfig(level=logging.DEBUG, format=colorlog_format)
+
+with gui_qt():
```
Comment (sofroniewn) [inline examples/node_tree.py]: I know you want us to move away from the gui_qt syntax long term, is that possible here too?

File: napari/_qt/tree/_tests/test_qt_tree_model.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    _assert_models_synced(root, tree_model)
+    root.move(3, 1)
+    _assert_models_synced(root, tree_model)
+
+
+@pytest.mark.parametrize('sources, dest, expectation', POS_INDICES)
+def test_nested_move_multiple(sources, dest, expectation):
+    """Test that models stay in sync with complicated moves.
+
+    This uses mimeData to simulate drag/drop operations.
```
Comment (sofroniewn) [inline napari/_qt/tree/_tests/test_qt_tree_model.py:65]: very nice

File: napari/_qt/tree/qt_tree_model.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+            logger.debug(
+                f"dropMimeData: indices {moving_indices} ➡ {dest_idx}"
+            )
+
+            if len(moving_indices) == 1:
+                self._root.move(moving_indices[0], dest_idx)
+            else:
+                self._root.move_multiple(moving_indices, dest_idx)
+
+        # XXX: returning False seems to go against the Qt docs instructions, but
```
Comment (sofroniewn) [inline napari/_qt/tree/qt_tree_model.py]: Does this still need to be addressed or is this determined now?

File: napari/_qt/tree/qt_tree_view.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+from qtpy.QtWidgets import QAbstractItemView, QTreeView, QWidget
+
+from ...utils.tree.group import Group
+from .qt_tree_model import QtNodeTreeModel
+
+
+class QtNodeTreeView(QTreeView):
```
Comment (sofroniewn) [inline napari/_qt/tree/qt_tree_view.py:7]: So cool how simple this is

**Review by sofroniewn — APPROVED** [review]
This looks great @tlambert03 I love how simple the QTreeView. The QTreeModel is very nice too - it really is the better way of coding this instead of the widgets.

I left a couple minor comments on the PR. I noticed a couple "ToDo" still left, not sure if you want to revisit those before merge or if this is good to go from your point of view. I didn't see any red flags so I was fine.

I'll look at #2261 next, but I wonder if it might make sense to have one more generic PR that makes a `SelectableEventedTree` after this, following convos we've been having in #2271 and elsewhere.

When this is all done, I would like us to consider coming back and making a `QListView` and `QListModel` for our EventedList as I do think us and plugin developers might want to use that again.

File: napari/_qt/tree/qt_tree_model.py (hunk 3 above)
Comment (tlambert03) [reply napari/_qt/tree/qt_tree_model.py]: fixed in the next pr... just backported to this one.

**Comment by tlambert03** [comment]
> I noticed a couple "ToDo" still left

cleaned up... ready to go!

> I'll look at #2261 next, but I wonder if it might make sense to have one more generic PR that makes a SelectableEventedTree after this, following convos we've been having in #2271 and elsewhere.

will have a look.

> When this is all done, I would like us to consider coming back and making a QListView and QListModel for our EventedList as I do think us and plugin developers might want to use that again.

I agree, those would be nice!

**Review by sofroniewn — COMMENTED** [review]
Ok, thanks for those fixes, changes here look good to me too. I'm happy to merge now too, and then we can press on as discussed. I'll leave this open till tomorrow morning unless i hear otherwise from @jni

File: examples/node_tree.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+from napari.qt import get_app, run
```
Comment (jni) [inline examples/node_tree.py]: Why is `get_app()` needed? I also would prefer normalising on `napari.run()` rather than bare `run()` calls...

Other than that, a very cool example!

Comment (jni) [reply examples/node_tree.py]: Or is this because there's nothing napari related in this example? In which case, maybe we want to somehow de-emphasise it, e.g. rename to `_node_tree.py` or `examples/_dev/node_tree.py`?

File: napari/utils/tree/node.py [hunk 6]
```diff
@@ -39,25 +37,25 @@ def is_group(self) -> bool:
         """
         return False
 
-    def index_in_parent(self) -> int | None:
+    def index_in_parent(self) -> Optional[int]:
```
Comment (jni) [inline napari/utils/tree/node.py:40]: Why these changes?

**Review by jni — APPROVED** [review]
@tlambert03 I just left minor comments/questions, which can be addressed or ignored as you see fit. =)

File: examples/node_tree.py (hunk 5 above)
Comment (tlambert03) [reply examples/node_tree.py]: I like the dev folder idea,  there is indeed nothing napari in here.  get_app is required because TreeView needs an app before instantiation (we don't do the auto app creation anywhere but viewer.window).  I'll change it to use the qt app creation directly to make that clearer, after moving to dev folder

File: napari/utils/tree/node.py (hunk 6 above)
Comment (tlambert03) [reply napari/utils/tree/node.py:40]: Well, @Carreau pointed out in another PR that that syntax "works" but breaks the ability to use `get_type_hints`.  So, I'm refraining for now.  We can still use it later if we choose, and it won't be hard to change back... but I'm punting on the decision to break `get_type_hints` in order to gain that nice syntax for now

File: examples/node_tree.py (hunk 5 above)
Comment (tlambert03) [reply examples/node_tree.py]: fixed, moved to `examples/dev`, and provided more explanation in the module docstring.

**Comment by sofroniewn** [comment]
I like the `examples/dev`, that might be useful for other examples too, we can think about moving more over there gradually too! Merging now, thanks for taking a look juan
